### ISSUE TITLE
[Feature] More pooling layers

### ISSUE THREAD

**Issue by NeptuneIsTheBest** [ml-explore/mlx#400]
Hello, when I looked at the PR #357 I found that there are only a few basic pooling layer implementations such as MaxPooling and AvgPooling. Will we add an adaptive pooling layer implementation similar to that in PyTorch?

**Comment by gboduljak** [comment]
@NeptuneIsTheBest I am willing to implement this after we get some feedback on MaxPooling and AvgPooling. I would like to add these in a separate PR.

**Comment by gboduljak** [comment]
> Hello, when I looked at the PR #357 I found that there are only a few basic pooling layer implementations such as MaxPooling and AvgPooling. Will we add an adaptive pooling layer implementation similar to that in PyTorch?

Meanwhile, if you need this feature earlier, you can implement these easily using ``mx.max`` and ``mx.mean``.

**Comment by NeptuneIsTheBest** [comment]
> @NeptuneIsTheBest This is a good suggestion. I am willing to implement this after we get some feedback on MaxPooling and AvgPooling. I would like to add these in a separate PR. @awni @angeloskath should we address this within the existing PR #357?

I'm looking to port my model (based on MobileNet-v3-Small) from PyTorch to mlX, so I'm looking to see if there are plans for AdaptiveAvgPooling.

> > Hello, when I looked at the PR #357 I found that there are only a few basic pooling layer implementations such as MaxPooling and AvgPooling. Will we add an adaptive pooling layer implementation similar to that in PyTorch?
> 
> Meanwhile, if you need this feature earlier, you can implement these easily using `mx.max` and `mx.mean`.

I'll try to implement an AdaptiveAvgPooling myself, thanks for the advice! Hope to see these Pooling officially released soon! :)

**Comment by gboduljak** [comment]
> > @NeptuneIsTheBest This is a good suggestion. I am willing to implement this after we get some feedback on MaxPooling and AvgPooling. I would like to add these in a separate PR. @awni @angeloskath should we address this within the existing PR #357?
> 
> I'm looking to port my model (based on MobileNet-v3-Small) from PyTorch to mlX, so I'm looking to see if there are plans for AdaptiveAvgPooling.
> 
> > > Hello, when I looked at the PR #357 I found that there are only a few basic pooling layer implementations such as MaxPooling and AvgPooling. Will we add an adaptive pooling layer implementation similar to that in PyTorch?
> > 
> > 
> > Meanwhile, if you need this feature earlier, you can implement these easily using `mx.max` and `mx.mean`.
> 
> I will try to implement an AdaptiveAvgPooling myself first, thanks for your suggestions! Hope to see these Pooling officially released soon! :)

@NeptuneIsTheBest In [this implementation of MobileNetV3](https://github.com/pytorch/vision/blob/main/torchvision/models/mobilenetv3.py), adaptive pooling is used to implement global average pooling. This can be implemented without adaptive pooling, namely using ``mlx.mean(x, (1,2))``. The calculation assumes MLX channels-last convention (e.g. input is of the shape ``(B, H, W, C)``). For this, you do not need a full implementation of adaptive pooling.

**Comment by NeptuneIsTheBest** [comment]
@gboduljak You are right, but I actually use AdaptiveAvgPool2D and AdaptiveMaxPool2D in many places. And (as a consequence of not spending the money to upgrade to 36GB of unified memory) I have to constantly tweak the model to test whether I'm going to run into those memory issues (I've added too many things to it, like attention and feature fusion, etc.). From an ease of use and code readability perspective, maybe implementing one in its entirety would be a better option? :>

**Comment by gboduljak** [comment]
> @gboduljak You are right, but I actually use AdaptiveAvgPool2D and AdaptiveMaxPool2D in many places. And (as a consequence of not spending the money to upgrade to 36GB of unified memory) I have to constantly tweak the model to test whether I'm going to run into those memory issues (I've added too many things to it, like attention and feature fusion, etc.). From an ease of use and code readability perspective, maybe implementing one in its entirety would be a better option? :>

I think having a full implementation of all 'adaptive' pooling layers is useful. It might be worthy to wait until 'simple' pooling layers are implemented, because it is possible to use these to implement the 'adaptive' ones. My comment was only to give a quick temporary solution, until this is done.

**Comment by NeptuneIsTheBest** [comment]
@gboduljak I've been implementing this as a temporary solution based on the method mentioned in #25 and provided by you. :>This should get me through the period leading up to the official release.

**Comment by rounak** [comment]
As an ML beginner, would love it if there was support for AdaptiveAvgPool2d.

**Comment by zcbenz** [comment]
As commented in https://github.com/ml-explore/mlx/pull/2487#issuecomment-3191421118, we don't have plan to add these layers in core.
